hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/slowlog/SlowLogTableAccessor.java

```diff
@@ -0,0 +1,120 @@
+
+  private static final Random RANDOM = new Random();
+
+  private static void doPut(final Connection connection, final Put put)
+      throws IOException {
+    try (Table table = connection.getTable(TableName.SLOW_LOG_TABLE_NAME)) {
+      table.put(put);
+    }
+  }
+
+  /**
+   * Add slow/large log records to hbase:slowlog table
+   *
+   * @param slowLogPayload SlowLogPayload to process
+   * @param connection Connection to put data
+   */
+  public static void addSlowLogRecord(final TooSlowLog.SlowLogPayload slowLogPayload,
+      final Connection connection) {
+    final byte[] rowKey = getRowKey(slowLogPayload);
+    final Put put = new Put(rowKey)
+      .setDurability(Durability.SKIP_WAL)
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("call_details"),
+        Bytes.toBytes(slowLogPayload.getCallDetails()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("client_address"),
+        Bytes.toBytes(slowLogPayload.getClientAddress()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("method_name"),
+        Bytes.toBytes(slowLogPayload.getMethodName()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("param"),
+        Bytes.toBytes(slowLogPayload.getParam()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("processing_time"),
+        Bytes.toBytes(slowLogPayload.getProcessingTime()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("queue_time"),
+        Bytes.toBytes(slowLogPayload.getQueueTime()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("region_name"),
+        Bytes.toBytes(slowLogPayload.getRegionName()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("response_size"),
+        Bytes.toBytes(slowLogPayload.getResponseSize()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("server_class"),
+        Bytes.toBytes(slowLogPayload.getServerClass()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("start_time"),
+        Bytes.toBytes(slowLogPayload.getStartTime()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("type"),
+        Bytes.toBytes(slowLogPayload.getType().name()))
+      .addColumn(HConstants.SLOWLOG_INFO_FAMILY, Bytes.toBytes("username"),
+        Bytes.toBytes(slowLogPayload.getUserName()));
+    try {
+      doPut(connection, put);
+    } catch (IOException e) {
+      LOG.error("Failed to add slow/large log record to hbase:slowlog table for region: {}",
+        slowLogPayload.getRegionName(), e);
+    }
+  }
+
+  /**
+   * Create rowKey: currentTimeMillis APPEND slowLogPayload.hashcode
+   * Scan on slowlog table should keep records with sorted order of time, however records
+   * added at the very same time (currentTimeMillis) could be in random order.
+   *
+   * @param slowLogPayload SlowLogPayload to process
+   * @return rowKey byte[]
+   */
+  private static byte[] getRowKey(final TooSlowLog.SlowLogPayload slowLogPayload) {
```

Comment:
Will this genereate hot spot on the region?

Reply:
Yes it can have hotsopts, but we are not expecting the system to be too slow for longer time. And also, when users scan the system table, they would expect to see slow RPC logs in increasing order of time.

Reply:
Added chore to insert records every 10 min, hence we are not much worried about hotspot.

Reply:
In a 10 min window we run the chore - but if there are slow sync happening during that 10 min window - we accumulate it in memory ?

Reply:
Ok I saw this public static final int DEFAULT_SLOW_LOG_RING_BUFFER_SIZE = 256;
So at max it is only 256 elements that we store in that queue. Rest will be removed anyway. At max we have 256 put list at any point of time. Right?

Reply:
Just one Q - this system table if not coming online due to some assignment issues -  we don't bother right? Say if the table is not online and while doing the puts we don't get the table - we should not do a WARN for certain time and then stop the connection fetching operatoin itself. Something like file based bucket cache - if there is afailure in writing to cache we keep retrying and if the error count exceeds a threshold we will WARN and no longer cache the data.

Reply:
`DEFAULT_SLOW_LOG_RING_BUFFER_SIZE = 256` this is for in-memory ring buffer only. For 10 min window, yes we will keep records in memory until we can insert them together by cron. And the size for the memory queue is 1000. The code is here:
```
    if (isSlowLogTableEnabled) {
      EvictingQueue<SlowLogPayload> evictingQueueForTable = EvictingQueue.create(
        SYS_TABLE_QUEUE_SIZE);
      queueForSysTable = Queues.synchronizedQueue(evictingQueueForTable);
    } else {
      queueForSysTable = null;
    }
```

> this system table if not coming online due to some assignment issues - we don't bother right?

That is true, and yes warn should be present, which is provided here:

```
    } catch (Exception e) {
      LOG.warn("Failed to add slow/large log records to hbase:slowlog table.", e);
    }
```

Retries and timeout is also defined here:
```
      Configuration conf = new Configuration(configuration);
      // rpc timeout: 20s
      conf.setInt(HConstants.HBASE_RPC_TIMEOUT_KEY, 20000);
      // retry count: 5
      conf.setInt(HConstants.HBASE_CLIENT_RETRIES_NUMBER, 5);
      conf.setInt(HConstants.HBASE_CLIENT_SERVERSIDE_RETRIES_MULTIPLIER, 1);
      connection = ConnectionFactory.createConnection(conf);
```